### PR TITLE
Decorated graphs

### PR DESCRIPTION
**Pull request by JoeyT1994**
This PR, which addresses #36, adds functionality for decorating graphs by adding additional graphs on their edges or vertices.

Two functions are provided 

`decorate_graph_edges(g::NamedGraph; edge_map::Function=e -> named_grid((1,)))`

and 

`decorate_graph_vertices(g::NamedGraph; vertex_map::Function=v -> named_grid((1,)))`

Currently there are defaults for which vertices of the decoration are connected to the edges/vertices original graph but we will add keyword arguments for this in the future.

Testing is included. This PR also fixed an ambiguity issue with function `a_star` when passed a `NamedGraph` with Integer labelling.

### CODE REVIEW

File: src/abstractnamedgraph.jl [hunk 1]
```diff
@@ -295,6 +295,29 @@ function a_star(
   return parent_edges_to_edges(graph, parent_shortest_path)
 end
 
+function a_star(
+  graph::AbstractNamedGraph,
+  source,
+  destination,
+  distmx=weights(graph),
+  heuristic::Function=(v -> zero(eltype(distmx))),
+  edgetype_to_return=edgetype(graph),
```
Comment (mtfishman) [inline src/abstractnamedgraph.jl:304]: We could use `args...` here so we don't have to list the arguments explicitly (besides the ones involved in the ambiguity error). Same for the more narrow `a_star` below.

Comment (mtfishman) [reply src/abstractnamedgraph.jl:304]: Also, could you add a short test for `a_star` with a named graph with integer vertices?

**Comment by mtfishman** [comment]
For the current version of `vertex_map`, aren't the edges connecting to the decorated vertex essentially random since the ordering that `neighbors` outputs is an internal detail? Seems like until we decide how to specify those new edges, it's not so useful.

**Comment by mtfishman** [comment]
Maybe `vertex_map` could be a function that accepts a list of vertices, where the first one is the vertex and the rest are the neighbors of that vertex. Then the user is expected to add edges between their new vertices in the vertex decoration and the neighbor vertices. If those aren't added, then those edges could be added (essentially) randomly as you do it now.

File: src/abstractnamedgraph.jl (hunk 1 above)
Comment (JoeyT1994) [reply src/abstractnamedgraph.jl:304]: Sure I will do that

**Comment by JoeyT1994** [comment]
> For the current version of `vertex_map`, aren't the edges connecting to the decorated vertex essentially random since the ordering that `neighbors` outputs is an internal detail? Seems like until we decide how to specify those new edges, it's not so useful.

Currently the vertex `vd` which is being decorated is removed and replaced with the decoration graph `gd.` The edges which originally connected `vd` to its neighbors in the undecorated graph `vd_neighbors` are reconnected to the first vertex of `gd`? So there is nothing random going on but obviously ideally the user should decide which vertices of `gd` should connect up to which members of `vd_neighbors` ?

**Comment by mtfishman** [comment]
Oh I see, I was picturing that each vertex neighbor would get connected to unique vertices of the decorated graph getting inserted onto the vertex, that's why I thought it would be random to some extent. In that case, I agree it isn't random since the user can choose which is the first vertex of the decorated vertex graph. In that case it seems fine to leave it for future work to give more control to the users over how the edges to the neighbors get specified.

**Pull request merged**